tests/conftest.py

```diff
@@ -119,11 +118,119 @@ def generate_tests(self, smoke_test_data, random_test_data=[]):
         _: list
             Tests.
         """
-        smoke_tests = [
-            pytest.param(*data.values(), marks=smoke) for data in smoke_test_data
-        ]
-        random_tests = [pytest.param(*data, marks=random) for data in random_test_data]
-        return smoke_tests + random_tests
+        tests = []
+        if smoke_test_data:
+            smoke_tests = [
+                pytest.param(*data.values(), marks=pytest.mark.smoke)
+                for data in smoke_test_data
+            ]
+            tests += smoke_tests
+        if random_test_data:
+            random_tests = [
+                pytest.param(*data.values(), marks=pytest.mark.random)
+                if isinstance(data, dict)
+                else pytest.param(*data, marks=pytest.mark.random)
+                for data in random_test_data
+            ]
+            tests += random_tests
+        return tests
+
+    def _log_exp_composition_data(
+        self, space, num_samples=100, max_n=10, num_n=5, **kwargs
+    ):
+        """Generate Data that checks for log and exp are inverses. Specifically
+
+            :math: `Exp_{base_point}(Log_{base_point}(point)) = point`
+
+
+        Parameters
+        ----------
+        space : cls
+            Manifold class that upon which metric is present.
+        max_n : int
+            Range of 'n' to generated.
+            Optional, default: 20
+        num_n : int
+            Number of 'n' to be generated.
+            Optional, default: 5
+        num_samples : int
```

Comment:
it's n_samples in the codebase (following scikit learn's API https://github.com/scikit-learn/scikit-learn/blob/9816b35d05e139f1fcc1a5541a1398205280d75a/sklearn/cluster/_agglomerative.py#L70): either keep it `n_samples` or change to `num_samples` everywhere.